src/ClientPool.cpp

```diff
@@ -0,0 +1,156 @@
+//#include <execution>
+#include <thread>
+#include "ClientPool.h"
+
+namespace douban {
+namespace mc {
+
+// default max of 4 clients to match memcached's default of 4 worker threads
+ClientPool::ClientPool(): m_initial_clients(1), m_max_clients(4), m_max_growth(4) {
+  memset(m_opt_changed, false, sizeof m_opt_changed);
+  memset(m_opt_value, 0, sizeof m_opt_value);
+}
+
+ClientPool::~ClientPool() {
+}
+
+void ClientPool::config(config_options_t opt, int val) {
+  std::lock_guard config_pool(m_pool_lock);
+  if (opt < CLIENT_CONFIG_OPTION_COUNT) {
+    m_opt_changed[opt] = true;
+    m_opt_value[opt] = val;
+    for (auto &client : m_clients) {
+      client.c.config(opt, val);
+    }
+    return;
+  }
+  std::unique_lock initializing(m_acquiring_growth);
+  switch (opt) {
+    case CFG_INITIAL_CLIENTS:
+      m_initial_clients = val;
+      if (m_initial_clients > m_max_clients) {
+        m_max_clients = m_initial_clients;
+      }
+      if (m_clients.size() < m_initial_clients) {
+        growPool(m_initial_clients);
+      }
+      break;
+    case CFG_MAX_CLIENTS:
+      m_max_clients = val;
+      break;
+    case CFG_MAX_GROWTH:
+      m_max_growth = val;
+      break;
+    default:
+      break;
+  }
+}
+
+int ClientPool::init(const char* const * hosts, const uint32_t* ports,
+                     const size_t n, const char* const * aliases) {
+  updateServers(hosts, ports, n, aliases);
+  std::unique_lock initializing(m_acquiring_growth);
+  std::lock_guard config_pool(m_pool_lock);
+  return growPool(m_initial_clients);
+}
+
+int ClientPool::updateServers(const char* const* hosts, const uint32_t* ports,
+                              const size_t n, const char* const* aliases) {
+  std::lock_guard updating_clients(m_pool_lock);
+  duplicate_strings(hosts, n, m_hosts_data, m_hosts);
+  duplicate_strings(aliases, n, m_aliases_data, m_aliases);
+
+  m_ports.resize(n);
+  std::copy(ports, ports + n, m_ports.begin());
+
+  std::atomic<int> rv = 0;
+  std::lock_guard<std::mutex> updating(m_fifo_access);
+  std::for_each(irange(0), irange(m_clients.size()),
+  //std::for_each(std::execution::par_unseq, irange(0), irange(m_clients.size()),
```

Comment:
The loops for initializing and growing the client pool (doubles in size each time it's overrun) were written using `for_each` in order to support parallelized initialization of the client objects, but at the scales I tested it at, the parallelization didn't make a difference, thus the commented out execution policy.